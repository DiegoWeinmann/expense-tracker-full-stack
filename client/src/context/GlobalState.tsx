import React, { createContext, useReducer } from 'react';

import { GlobalState, GlobalStateReducer, GlobalStateDispatch } from '../types';
import { AppReducer } from '../reducers/AppReducer';

const initialState: GlobalState = {
  transactions: [],
};

export const GlobalContext = createContext<GlobalState>(initialState);
export const GlobalContextDispatch = createContext<GlobalStateDispatch>({
  dispatch: () => {},
});

export const GlobalProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer<GlobalStateReducer>(
    AppReducer,
    initialState
  );

  return (
    <GlobalContext.Provider value={{ transactions: state.transactions }}>
      <GlobalContextDispatch.Provider value={{ dispatch }}>
        {children}
      </GlobalContextDispatch.Provider>
    </GlobalContext.Provider>
  );
};
