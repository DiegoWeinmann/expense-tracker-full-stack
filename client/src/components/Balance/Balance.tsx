import React, { useContext } from 'react';
import { GlobalContext } from '../../context/GlobalState';

export const Balance: React.FC = () => {
  const { transactions } = useContext(GlobalContext);
  const total = transactions
    .reduce((acc, transaction) => acc + transaction.amount, 0)
    .toFixed(2);

  return (
    <>
      <h4>Your Balance</h4>
      <h1 className="balance">${total}</h1>
    </>
  );
};
