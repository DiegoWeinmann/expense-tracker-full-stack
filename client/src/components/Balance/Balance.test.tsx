import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { Balance } from './Balance';
import { GlobalProvider } from '../../context/GlobalState';

afterEach(cleanup);

const renderUI = () => {
  return render(
    <GlobalProvider>
      <Balance />
    </GlobalProvider>
  );
};

describe('<Balance/>', () => {
  it('renders without crashing.', () => {
    const { container } = renderUI();
    expect(container).toBeInTheDocument();
  });

  it('matches the snapshot.', () => {
    const { asFragment } = renderUI();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders an initial total of $0.00', () => {
    const { getByText } = renderUI();
    expect(getByText('$0.00')).toBeInTheDocument();
  });
});
