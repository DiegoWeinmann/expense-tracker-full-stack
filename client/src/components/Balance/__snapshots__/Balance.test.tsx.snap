// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Balance/> matches the snapshot. 1`] = `
<DocumentFragment>
  <h4>
    Your Balance
  </h4>
  <h1
    class="balance"
  >
    $0.00
  </h1>
</DocumentFragment>
`;
